idea-plugin/src/com/reason/ide/annotations/ErrorAnnotator.java

```diff
@@ -1,113 +1,229 @@
+    // It'll be used by bsc for a temporary compilation
+    Path tempFilePath = Paths.get(m_compilationDirectory.getPath(), sourceFile.getName());
+    try {
+      Files.write(tempFilePath, psiFile.getText().getBytes());
+    } catch (IOException e) {
+      LOG.error("Failed to write to temporary file.", e);
+      return null;
+    }
+    LOG.debug("Wrote contents to temporary file.", tempFilePath);
+
+    File cmtFile = new File(m_compilationDirectory, sourceFile.getNameWithoutExtension() + ".cmt");
+
+    List<String> arguments = new ArrayList<>();
+    arguments.add("-bs-super-errors");
+    arguments.add("-color");
+    arguments.add("never");
+    arguments.addAll(ninja.getPkgFlags());
+    arguments.addAll(ninja.getBscFlags());
+    for (String ppxPath : ninja.getPpxIncludes()) {
+      arguments.add("-ppx");
+      arguments.add(ppxPath);
+    }
+    if (!namespace.isEmpty()) {
+      arguments.add("-bs-ns");
+      arguments.add(namespace);
+    }
+    if (jsxVersion != null) {
+      arguments.add("-bs-jsx");
+      arguments.add(jsxVersion);
+    }
+    for (String bscInclude : ninja.getIncludes()) {
+      arguments.add("-I");
+      arguments.add(bscInclude);
+    }
+    arguments.add("-o");
+    arguments.add(cmtFile.getAbsolutePath());
+    arguments.add("-bin-annot");
+    arguments.add(tempFilePath.toString());
+
+    return new InitialInfo(psiFile, cmtFile, editor, arguments);
   }
 
   @Nullable
   @Override
-  public Collection<BsbErrorAnnotation> doAnnotate(
-      @NotNull Collection<BsbErrorAnnotation> collectedInfo) {
-    return collectedInfo.isEmpty() ? null : collectedInfo;
+  public AnnotationResult doAnnotate(@Nullable InitialInfo initialInfo) {
+    if (initialInfo == null) {
+      LOG.warn("Unable to annotate file. Annotator not ready.");
+      return null;
+    }
+
+    PsiFile sourcePsiFile = Objects.requireNonNull(initialInfo.sourcePsiFile);
+    Project project = sourcePsiFile.getProject();
+    VirtualFile sourceFile = sourcePsiFile.getVirtualFile();
+
+    long compilationStartTime = System.currentTimeMillis();
+
+    BscProcess bscProcess = BscProcess.getInstance(project);
+    BscProcessListener bscListener = new BscProcessListener();
+
+    Integer exitCode = bscProcess.run(sourceFile, initialInfo.arguments, bscListener);
+    if (exitCode == null) {
+      LOG.error("Something went wrong when running bsc.exe");
+      return null;
+    }
+    LOG.trace("Compilation done in " + (System.currentTimeMillis() - compilationStartTime) + "ms");
+
+    if (exitCode == 0) {
+      ApplicationManager.getApplication()
+          .invokeLater(
+              () -> {
+                if (!project.isDisposed()) {
+                  PsiFile psiFile = PsiManager.getInstance(project).findFile(sourceFile);
+                  if (psiFile != null) {
+                    LOG.trace("Restart daemon code analyzer for " + psiFile);
+                    DaemonCodeAnalyzer.getInstance(project).restart(psiFile);
```

Comment:
Is that really needed ? I used that code to force a restart of the external annotator. 
But here we are already running that annotator